### ISSUE TITLE
1st review

### ISSUE THREAD

**Issue by d-qql** [ElementSi/SoLE#1]
1) Общие замечания по тестам и исходникам, описанные в беседе, общие замечания по конструкторам.

2)
```
std::vector<double> SolveTridiagonalMatrix(unsigned int n,
                                           const TridiagonalMatrix& A,
                                           const std::vector<double>&& f);
```

С какой целью здесь принимается rvalue-reference?

**Comment by ElementSi** [comment]
1. По поводу исходников и тестов: "Необходимо разнести исходники и тесты по своим директориям src и tests с соответствующими изменениями в CMakeLists.txt: создать субдиректории, в симейке исходников генерировать не исполняемый файл, а бибилотеку, в симейке тестов генерировать исполняемые файлы, линкуя созданную библиотеку. Также напоминаю про функцию file(GLOB_RECURSE... ) чтобы не прописывать все исходники руками" - таков был общий комментарий.

Насколько я вижу, у меня всё соответствует описанию. Единственное - исходники лежали в субдиректории "tridiagonal_matrix" в папке проекта, я подвинул их глубже, добавив папку source. Однако CMakeLists.txt для исходников всё равно же придётся оставить в корне? Иначе ошибка ```CMakeLists.txt not found in D:\code\SoLE Select CMakeLists.txt```.

По поводу конструкторов - я не следую правилам, у меня нет ни конструкторов копирования, ни перемещения, ни каких-то похожих вещей. Но про них много было написано, может быть я что-то пропустил, а чего-то не понял

2. rvalue reference с целью, чтобы можно было сразу писать циферки туда, не создавая отдельно вектор соответствующий. Я это не использую, но у матриц в конструкторе такая функция заложена - почему бы и здесь её не предоставить?

**Comment by d-qql** [comment]
> Однако CMakeLists.txt для исходников всё равно же придётся оставить в корне? Иначе ошибка `CMakeLists.txt not found in D:\code\SoLE Select CMakeLists.txt.`

Нет, необходимо в основном симейке сделать сабдирректорию, ссылающуюся на папку исходников, как это сделано для тестов, и в ней создать еще один симейк лист, который будет генерировать библиотеку

**Comment by ElementSi** [comment]
> > Однако CMakeLists.txt для исходников всё равно же придётся оставить в корне? Иначе ошибка `CMakeLists.txt not found in D:\code\SoLE Select CMakeLists.txt.`
> 
> Нет, необходимо в основном симейке сделать сабдирректорию, ссылающуюся на папку исходников, как это сделано для тестов, и в ней создать еще один симейк лист, который будет генерировать библиотеку

Окей, такая логика понятна. В остальном всё ок?

**Comment by d-qql** [comment]
Нет, по поводу rvalue - ссылки все-таки есть замечания, есть предположение, что ты не понимаешь, что это такое. Допустим есть такой код:

```cpp
void foo(std::vector<double>&& a){
...
}
```

Он означает, что можно вызвать foo таким образом: foo({1, 2, 3, 4}),

Но при этом не будет проблем с тем, чтобы внутри foo производить действия над аргументом, а именно писать в него.

У тебя же две с половиной проблемы в коде:
1) Везде, где ты написал rvalue ссылки ты не пользуешься тем, что они предоставляют никогда, т.е. записью в аргумент, а простой вызов foo({1, 2, 3, 4}) возможен и просто по константной ссылке.
2) Ты раздул код в два раза, просто потому что тебе захотелось rvalue, которое не будет вызвано никогда, при том код абсолютно одинаков в обеих функциях.
3) Даже если бы ты хотел пользоваться тем, что предоставляет тебе прием rvalue, у тебя бы ничего не вышло, потому что ты повесил на него const

**Comment by d-qql** [comment]
> По поводу конструкторов - я не следую правилам, у меня нет ни конструкторов копирования, ни перемещения, ни каких-то похожих вещей. Но про них много было написано, может быть я что-то пропустил, а чего-то не понял

Ты упустил мою притчу про констовость аргументов
